### DISCUSSION TITLE
Contribution: Solax Parallel Mode Development and Testing

### DISCUSSION THREAD

**Discussion by fxstein** [wills106/homeassistant-solax-modbus#1053] [General]
Hi,

I'd like to contribute some of the Solax parallel mode learnings and fixes. I have access to multiple Solax installs including a 3 x Gen3 X4 15kW (45kW total with 68kW battery). I could not find a contributor guide and before I start hacking away I wanted to reach out to ask for help to get me set up. I have been running these systems for 18+ months and finally got them very stable with proper Solax settings, cabling and even using the WLAN adapters for all the modbus communication with consistent and high availability.

First I'd like to get my private code fixes into the main line of code, but also discuss some potential changes for larger scale installs. I'd be also happy to help with parallel mode testing and debugging going forward as well as help with some of the parallel mode Q&A and documentation.

Would appreciate your help to get me properly setup and started.

Best,

Oliver

**Comment by romankysely** [comment]
Hi I have 2x X3 Gen4 12kW + 46kW Battery in parallel mode 18 months with list of problematic topics and glitchies. I would like to be involved in this effort.
As Wills106 wrote, he does not have possibility to test parall mode unfrotunately so now we can contribute.

**Comment by wills106** [comment]
I don't really have any specific advice to get started, as I'm not a programmer by trade.
You're probably best running a Modbus proxy like mentioned [here](https://github.com/wills106/homeassistant-solax-modbus?tab=readme-ov-file#multiple-connections) then you can have a second instance of connecting to the Inverter as quite often they only allow a single Modbus connection. Then you can develop on a separate HA instance to your main setup, so there isn't a risk to messing your main setup up.

**Comment by fxstein** [comment]
I have added a PM skeleton as an early pull request here: https://github.com/wills106/homeassistant-solax-modbus/pull/1076

It fixes the worst issues when charging from the grid, but I am still working on correcting several of the calculations that are creating wrong numbers (like house load) when in PM setup.

**Comment by fxstein** [comment]
In addition I have opened a feature request for the support of the Solax DataHub 1000 as the primary RS485 gateway here: https://github.com/wills106/homeassistant-solax-modbus/issues/1078

This is related to the PM work I am doing as it would allow the use of the DataHub as the primary modbus gateway for all attached inverters and EV chargers.
